File: Python3/1 - 200/154 Find Minimum in Rotated Sorted Array II.py
# 40ms, 75.38%; 13.3MB, 5.63%
class Solution:
    def findMin(self, nums: list) -> int:
        if nums[0] == nums[-1]:
            j = len(nums) - 1
            while j > 0 and nums[j] == nums[-1]:
                j -= 1
            nums = nums[:1 + j]

        length = len(nums)
        if length == 1 or nums[0] < nums[-1]: return nums[0]
        if length == 2: return min(nums)
        if nums[1] < nums[0] and nums[1] > nums[-1]: return nums[-1]
        
        i, j = 0, length - 1
        while i < j:
            mid = (i + j) // 2
            if nums[mid] >= nums[0]:
                i = mid + 1
            else:
                j = mid
        return nums[i]

File: 1 - 200/154 Find Minimum in Rotated Sorted Array II.cc
// 4ms, 100.0%; 8.9MB, 31.25%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int findMin(vector<int> &nums) {
        if (nums[0] == nums.back()) {
            auto it = find_if(nums.rbegin(), nums.rend(), [&](int x) { return x != nums[0]; });
            if (it == nums.rend()) return nums[0];
            nums.resize(distance(it, nums.rend()));
        }

        int length = nums.size();
        if (length == 1 || nums[0] < nums.back()) return nums[0];
        if (length == 2) return min(nums[0], nums[1]);
        if (nums[1] < nums[0] && nums[1] > nums.back()) return nums.back();

        int i = 0, j = length - 1;
        while (i < j) {
            int mid = i + (j - i) / 2;
            if (nums[mid] >= nums[0]) {
                i = mid + 1;
            } else {
                j = mid;
            }
        }
        return nums[i];
    }
};
